### DISCUSSION TITLE
change the ManyButtonsSelector

### DISCUSSION THREAD

**Discussion by santta** [tttapa/Control-Surface#495] [Show and tell]
Hi, how do I change the ManyButtonsSelector.hpp so that the ManyButtonsSelectorLeds are sent every click by the PC?  Default settings the selector ignores repeated pressing of the button too.

**Comment by tttapa** [comment]
You could try something like this (untested):

```cpp
template <setting_t N, class Callback = EmptySelectorCallback>
class GenericCustomManyButtonsSelector : public GenericSelector<N, Callback> {
    using Parent = GenericSelector<N, Callback>;

  public:
    GenericCustomManyButtonsSelector(Selectable<N> &selectable,
                                     const Callback &callback,
                                     const Array<AH::Button, N> &buttons)
        : Parent{selectable, callback}, buttons(buttons) {}

    void begin() override {
        Parent::begin();
        for (AH::Button &b : buttons)
            b.begin();
    }

    void update() override {
        Parent::update();
        for (setting_t i = 0; i < N; i++)
            if (buttons[i].update() == AH::Button::Falling)
                this->set(i);
    }

  private:
    Array<AH::Button, N> buttons;
};
```

**Comment by santta** [comment]
I got an error in my sketch after changing the selector:

	In file included from C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Control_Surface.h:105:0,
                 from I:\АРДУИНО\FOOTSWITCH\14-1_test_2_kn_3\14-1_test_2_kn_3.ino:1:
	C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Selectors/ManyButtonsSelector.hpp:53:62: error: expected template-name before '<' token
	class ManyButtonsSelector : public GenericManyButtonsSelector<N> {
                                                              ^
	C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Selectors/ManyButtonsSelector.hpp:53:62: error: expected '{' before '<' token
	In file included from C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Control_Surface.h:113:0,
                 from I:\АРДУИНО\FOOTSWITCH\14-1_test_2_kn_3\14-1_test_2_kn_3.ino:1:
	C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Selectors/LEDs/SelectorLEDs.hpp:154:40: error: expected template-name before '<' token
     : public GenericManyButtonsSelector<N, SelectorLEDsCallback<N>> {
                                        ^	
	C:\Users\Admin\Documents\Arduino\libraries\Control-Surface-1.2.0\src/Selectors/LEDs/SelectorLEDs.hpp:154:40: error: expected '{' before '<' token
	14-1_test_2_kn_3:23:28: error: variable 'CS::ManyButtonsSelectorLEDs<4> programSelector' has initializer but incomplete type

	exit status 1
	variable 'CS::ManyButtonsSelectorLEDs<4> programSelector' has initializer but incomplete type






My code:
       #include <Control_Surface.h>
	
    // Create two MIDI interfaces
    USBMIDI_Interface usbmidi;
    HardwareSerialMIDI_Interface serialmidi = {Serial1, MIDI_BAUD};
 
    // Create a MIDI pipe factory to connect the MIDI interfaces to Control Surface
    BidirectionalMIDI_PipeFactory<2> pipes;

    //ПРЕСЕТЫ
    ProgramChanger<4> programChanger = {
      { 
        0, // А
        1, // B
        5, // B
        6, // B

      },
      CHANNEL_1, // MIDI канал
    };

    // Instantiate a selector that reads three buttons and controls the program changer
    ManyButtonsSelectorLEDs<4> programSelector = {
      programChanger,
      {6, 9, 4, 7}, // button pins
      {5, 8, 2, 3}, // led pins
    };


    Button pushbutton1 = {10};                       //Кнопка 1 Группа 2
    Button pushbutton2 = {15};                       //Кнопка 2 Группа 2
    Button pushbutton3 = {6};                        //Кнопка 1 Группа 1
    Button pushbutton4 = {9};                        //Кнопка 2 Группа 1

    int led1pin = 16;                                  //Диод 1 Группа 2
    int led2pin = A0;                                  //Диод 2 Группа 2

    bool btn1_mode = false;                           //variable that alternate after each push of button 1
    bool btn2_mode = false;                           //variable that alternate after each push of button 2
    bool btn3_mode = false;                           //variable that alternate after each push of button 3
    bool btn4_mode = false;                           //variable that alternate after each push of button 4


    void setup() {
      // Manually connect the MIDI interfaces to Control Surface
      Control_Surface | pipes | usbmidi;
      Control_Surface | pipes | serialmidi;
      // Initialize Control Surface _after_ connecting the interfaces

      Control_Surface.begin();
      pushbutton1.begin();                            //enables internal pull-up for button 1
      pushbutton2.begin();                            //enables internal pull-up for button 2
      pushbutton3.begin();                            //enables internal pull-up for button 3
      pushbutton4.begin();                            //enables internal pull-up for button 4
      pinMode(led1pin, OUTPUT);                       //set led 1 pin as output
      pinMode(led2pin, OUTPUT);                       //set led 2 pin as output
    }


    void loop() {
      Control_Surface.loop();   // Update the Control Surface
      pushbutton1.update();                           // read button 1 state
      pushbutton2.update();                           // read button 2 state
      pushbutton3.update();                           // read button 3 state
      pushbutton4.update();                           // read button 4 state

    //Кнопка 1 группы 1____________________________________________________________________________________________
      if (pushbutton3.getState() == Button::Falling){  //когда кнопка 1, группы 1 нажата
        btn1_mode = false;                        //change the state of the variable after each press 
        btn2_mode = false;                        //change the state of the variable after each press       
        if (btn1_mode == false){                        //if variable is true 
          //Control_Surface.sendCC(2,0);                 //посылается СС 2 со значением ВКЛ
          analogWrite(led1pin, 0);                   //Диод 1 группы 2 ВКЛ 
        }
        if (btn2_mode == false){                       //if variable is false 
        //Control_Surface.sendCC(2,127);                 //посылается СС 2 со значением ВЫКЛ
        analogWrite(led2pin, 0);                       //Диод 2 группы 2 ВЫКЛ  
       }
      } 
     
    //Кнопка 1 группы 1_________________________________________________________________________________________
      if (pushbutton4.getState() == Button::Falling){  //когда кнопка 2 группы 2 нажата
        btn1_mode = false;                        //change the state of the variable after each press 
        btn2_mode = false;                        //change the state of the variable after each press     
           
        if (btn1_mode == false){                        //if variable is true 
      //Control_Surface.sendCC(2,0);                 //посылается СС 2 со значением ВКЛ
         analogWrite(led1pin, 0);                   //Диод 1 группы 2 ВКЛ 
        }
        if (btn2_mode == false){                       //if variable is false 
        //Control_Surface.sendCC(2,127);                 //посылается СС 2 со значением ВЫКЛ
        analogWrite(led2pin, 0);                       //Диод 2 группы 2 ВЫКЛ  
       }
      }  
  
    //Кнопка 1 группы 2  _______________________________________________________________________________________________ 
      if (pushbutton1.getState() == Button::Falling){ //когда кнопка 1 группы 2 нажата
        btn1_mode = !btn1_mode;                       //change the state of the variable after each press 
           
        if (btn1_mode == true){                       //if variable is true 
          Control_Surface.sendCC(3,0);                //посылается СС 3 со значением ВКЛ
          analogWrite(led1pin, 255);                  //Диод 1 группы 2 ВКЛ 
        }
        if (btn1_mode == false){                      //if variable is false 
          Control_Surface.sendCC(3,127);              //посылается СС 3 со значением ВЫКЛ
          analogWrite(led1pin, 0);                    //Диод 1 группы 2 ВЫКЛ                
        }
      }
  
    //Кнопка 2 группы 2 _________________________________________________________________________________________________
      if (pushbutton2.getState() == Button::Falling){  //когда кнопка 2 группы 2 нажата
        btn2_mode = !btn2_mode;                        //change the state of the variable after each press 
           
        if (btn2_mode == true){                        //if variable is true 
          Control_Surface.sendCC(2,0);                 //посылается СС 2 со значением ВКЛ
          analogWrite(led2pin, 255);                   //Диод 2 группы 2 ВКЛ 
        }
        if (btn2_mode == false){                       //if variable is false 
        Control_Surface.sendCC(2,127);                 //посылается СС 2 со значением ВЫКЛ
        analogWrite(led2pin, 0);                       //Диод 2 группы 2 ВЫКЛ  
       }
      } 

    }

**Reply by tttapa** [reply]
Did you change the library source files? The code I posted goes in your sketch, you shouldn't edit the library files.

**Comment by santta** [comment]
Yes, I changed the code in the library files. But I returned everything to its original state.

Not sure where in the sketch should I insert your code? 

Tried it in different places, but your code did not change anything in the operation of the device.

**Reply by tttapa** [reply]
You have to use the new class instead of the old one. I'll see if I find the time to write an example tonight.

**Reply by tttapa** [reply]
```cpp
#include <Control_Surface.h>

template <setting_t N, class Callback = EmptySelectorCallback>
class GenericCustomManyButtonsSelector : public GenericSelector<N, Callback> {
    using Parent = GenericSelector<N, Callback>;

  public:
    GenericCustomManyButtonsSelector(Selectable<N> &selectable,
                                     const Callback &callback,
                                     const Array<AH::Button, N> &buttons)
      : Parent{selectable, callback}, buttons(buttons) {}

    void begin() override {
      Parent::begin();
      for (AH::Button &b : buttons)
        b.begin();
    }

    void update() override {
      Parent::update();
      for (setting_t i = 0; i < N; i++)
        if (buttons[i].update() == AH::Button::Falling)
          this->set(i);
    }

  private:
    Array<AH::Button, N> buttons;
};

template <setting_t N>
class CustomManyButtonsSelectorLEDs
  : public GenericCustomManyButtonsSelector<N, SelectorLEDsCallback<N>> {
  public:
    CustomManyButtonsSelectorLEDs(Selectable<N> &selectable,
                                  const Array<AH::Button, N> &buttons,
                                  const PinList<N> &ledPins)
      : GenericCustomManyButtonsSelector<N, SelectorLEDsCallback<N>> {
      selectable,
      {ledPins},
      buttons,
    } {}
};

ProgramChanger<4> programChanger = {
  { 
    0, // А
    1, // B
    5, // B
    6, // B
  },
  CHANNEL_1,
};

CustomManyButtonsSelectorLEDs<4> programSelector = {
  programChanger,
  {6, 9, 4, 7}, // button pins
  {5, 8, 2, 3}, // led pins
};

// You'll have to use the right interface for your application, this is just for testing
USBDebugMIDI_Interface midi;

void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```

**Comment by santta** [comment]
thank you, I will wait very much

**Comment by santta** [comment]
Everything works, thanks, Pieter.

I am sharing my sketch.

Made it for FootSwitch using amplitube 5.
So that two pedals can be used in the rig. And so that the settings for the indication of turning on the pedals are lost when switching presets in the program.    
    
     //СКЕТЧ для применения с педалями amplitube 5
     #include <Control_Surface.h>

     USBMIDI_Interface midi; // Подключение USB интерфейса

     //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
     template <setting_t N, class Callback = EmptySelectorCallback>
     class GenericCustomManyButtonsSelector : public GenericSelector<N, Callback> {
         using Parent = GenericSelector<N, Callback>;

       public:
         GenericCustomManyButtonsSelector(Selectable<N> &selectable,
                                     const Callback &callback,
                                     const Array<AH::Button, N> &buttons)
      : Parent{selectable, callback}, buttons(buttons) {}

         void begin() override {
      Parent::begin();
      for (AH::Button &b : buttons)
        b.begin();
     }

         void update() override {
      Parent::update();
      for (setting_t i = 0; i < N; i++)
        if (buttons[i].update() == AH::Button::Falling)
          this->set(i);
     }

       private:
         Array<AH::Button, N> buttons;
     };

     template <setting_t N>
     class CustomManyButtonsSelectorLEDs
       : public GenericCustomManyButtonsSelector<N, SelectorLEDsCallback<N>> {
       public:
         CustomManyButtonsSelectorLEDs(Selectable<N> &selectable,
                                  const Array<AH::Button, N> &buttons,
                                  const PinList<N> &ledPins)
      : GenericCustomManyButtonsSelector<N, SelectorLEDsCallback<N>> {
      selectable,
      {ledPins},
      buttons,
         } {}
     };
     //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
     ProgramChanger<2> programChanger = {
       { 
         0, // А
         1, // B

       },
       CHANNEL_1,
     };

     CustomManyButtonsSelectorLEDs<2> programSelector = {
       programChanger,
       {6, 9}, // button pins
       {5, 8}, // led pins
     };


     //кнопки с фиксацией+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
     Button pushbutton1 = {10};                       //Кнорка педали
     Button pushbutton2 = {15};                       //Кнорка педали
     Button pushbutton3 = {6};                        //Кнорка переключения пресета
     Button pushbutton4 = {9};                        //Кнорка переключения пресета

     int led1pin = 16;                                        //кнопка педали
     int led2pin = A0;                                        //кнопка педали

     bool btn1_mode = false;                           //variable that alternate after each push of button 1
     bool btn2_mode = false;                           //variable that alternate after each push of button 2
     //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++



     void setup() {
       Control_Surface.begin();
       pushbutton1.begin();                            //enables internal pull-up for button 1
       pushbutton2.begin();                            //enables internal pull-up for button 2
       pushbutton3.begin();                            //enables internal pull-up for button 3
       pushbutton4.begin();                            //enables internal pull-up for button 4
       pinMode(led1pin, OUTPUT);                //set led 1 pin as output
       pinMode(led2pin, OUTPUT);                 //set led 2 pin as output
     }


     void loop() {
       Control_Surface.loop();
       pushbutton1.update();                         // read button 1 state
       pushbutton2.update();                           // read button 2 state
       pushbutton3.update();                           // read button 3 state
       pushbutton4.update();                           // read button 4 state
  
     //кнопка педали 1 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
       if (pushbutton1.getState() == Button::Falling){       //
         btn1_mode = !btn1_mode;                                   // 
       
         if (btn1_mode == true){                                       //
           Control_Surface.sendCC(3,0);                             //
           analogWrite(led1pin, 255);                                 //
         }
         if (btn1_mode == false){                                      //
           Control_Surface.sendCC(3,127);                         //
           analogWrite(led1pin, 0);                                    //                
         }
       }
  
     //кнопка педали 2 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
       if (pushbutton2.getState() == Button::Falling){    //
         btn2_mode = !btn2_mode;                                //
       
         if (btn2_mode == true){                                   //
      Control_Surface.sendCC(2,0);                              //
      analogWrite(led2pin, 255);                                  // 
         }
         if (btn2_mode == false){                                 //
         Control_Surface.sendCC(2,127);                      //
         analogWrite(led2pin, 0);                                 //
        }
       } 

  
      //Кнопка переключения пресета 1 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
       if (pushbutton3.getState() == Button::Falling){       //
         btn1_mode = false;                                               // 
         btn2_mode = false;                                               //   
         if (btn1_mode == false){                                      //
      //Control_Surface.sendCC(2,0);                               //
      analogWrite(led1pin, 0);                                         //
     }
     if (btn2_mode == false){                                        //
     //Control_Surface.sendCC(2,127);                          //
     analogWrite(led2pin, 0);                                        //  
     }
     } 
 
     //Кнопка переключения пресета 1 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++
     if (pushbutton4.getState() == Button::Falling){    //
     btn1_mode = false;                                              //
     btn2_mode = false;                                             //   
       
     if (btn1_mode == false){                                    //
      //Control_Surface.sendCC(2,0);                         //
      analogWrite(led1pin, 0);                                   // 
     }
     if (btn2_mode == false){                                  //
     //Control_Surface.sendCC(2,127);                    //
     analogWrite(led2pin, 0);                                   //
     }
     }  
     }
